library/ssl_tls13_generic.c

```diff
@@ -504,7 +506,62 @@ int mbedtls_ssl_tls13_parse_certificate( mbedtls_ssl_context *ssl,
         extensions_len = MBEDTLS_GET_UINT16_BE( p, 0 );
         p += 2;
         MBEDTLS_SSL_CHK_BUF_READ_PTR( p, certificate_list_end, extensions_len );
-        p += extensions_len;
+
+        extensions_end = p + extensions_len;
+        extensions_present = MBEDTLS_SSL_EXT_NONE;
+
+        while( p < extensions_end )
+        {
+            unsigned int extension_type;
+            size_t extension_data_len;
+
+            /*
+            * struct {
+            *     ExtensionType extension_type; (2 bytes)
+            *     opaque extension_data<0..2^16-1>;
+            * } Extension;
+            */
+            MBEDTLS_SSL_CHK_BUF_READ_PTR( p, extensions_end, 4 );
+            extension_type = MBEDTLS_GET_UINT16_BE( p, 0 );
+            extension_data_len = MBEDTLS_GET_UINT16_BE( p, 2 );
+            p += 4;
+
+            MBEDTLS_SSL_CHK_BUF_READ_PTR( p, extensions_end, extension_data_len );
+
+            /* RFC 8446 page 35
+             *
+             * If an implementation receives an extension which it recognizes and
+             * which is not specified for the message in which it appears, it MUST
+             * abort the handshake with an "illegal_parameter" alert.
+             */
+            extensions_present |= mbedtls_tls13_get_extension_mask( extension_type );
+            MBEDTLS_SSL_DEBUG_MSG( 3,
+                        ( "encrypted extensions : received %s(%u) extension",
+                        mbedtls_tls13_get_extension_name( extension_type ),
+                        extension_type ) );
+            if( ( extensions_present & MBEDTLS_SSL_TLS1_3_ALLOWED_EXTS_OF_CT ) == 0 )
```

Comment:
We should also check that we sent the extension in the ClientHello or CertificateRequest message.

Reply:
It has been re-written.